Prometheus: Move existing query logic to new buffered package

toddtreece opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

-->

**What this PR does / why we need it**:

This is the first step in a series of PRs that will use the prometheus JSON streaming parser added by https://github.com/grafana/grafana/pull/44520.

The current plan is:
1. Move existing logic to `buffered` package (this PR) 
2. Add separate `querydata` package that uses a standard HTTP client instead of the prometheus go client and the streaming JSON response parser (https://github.com/grafana/grafana/pull/48477)
3. Remove `buffered` package & feature flag (future PR once things have stabilized)

**Which issue(s) this PR fixes**:

<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes #

**Special notes for your reviewer**:

toddtreece committed:
Prometheus: Move existing query logic to new package

toddtreece committed:
fix tests

toddtreece committed:
fix lint error

ryantxu commented on pkg/tsdb/prometheus/buffered/types.go:
```diff
@@ -0,0 +1,49 @@
+
+type clientGetter func(map[string]string) (apiv1.API, error)
+
+type PrometheusQuery struct {
```
how does this relate to:
pkg/tsdb/prometheus/types.go

toddtreece replied:
good catch. i missed removing `pkg/tsdb/prometheus/types.go`

toddtreece committed:
delete unused types file

ryantxu commented:
Are there any behavioral changes here?  It looks like tracer is passed around differently, but I can not tell if it adds up to the same thing :)

Given the changes required for the raw json parsing, do we *need* a different version of testdata?  and types.go?

toddtreece committed:
move testdata back to original location

toddtreece commented:
> Are there any behavioral changes here? It looks like tracer is passed around differently, but I can not tell if it adds up to the same thing :)

@ryantxu there shouldn't be any behavioral changes with this one. all of the minor changes are related to passing around dependencies a little differently due to the move to a separate package

> Given the changes required for the raw json parsing, do we need a different version of testdata? and types.go?

good call on testdata, i moved that back and updated the path in the test. i think there are enough changes to `types.go` in the next PR to keep it as a separate file in `buffered`

ryantxu commented:
I like the approach -- I think we *could* find a way to do this without exposing public functions, but also don't think it matters enough to worry about as part of a migration path.

For this PR, the @grafana/observability-metrics needs to feel comfortable that this does not introduce any real changes, just structures things for easier comparison in the *very near* future.

gabor reviewed (approved):
LGTM!
